cpp_tests/test_Test.cpp

```diff
@@ -0,0 +1,417 @@
+	arraysEqual<Test_logevent_arraysC>(expected_data, &data);
+
+	// at this point there should be nothing on the queue
+	REQUIRE(remote->getNextSample_logevent_arrays(&data) == SAL__NO_UPDATES);
+	REQUIRE(remote->getSample_logevent_arrays(&data) == SAL__NO_UPDATES);
+    }
+
+    /**
+     * Write several messages and make sure gettting the newest returns that and flushes the queue.
+     */
+    SECTION("TelGetNewest") {
+	remote->salTelemetrySub((char *)"Test_arrays");
+	controller->salTelemetryPub((char *)"Test_arrays");
+
+	Test_arraysC dataList[numLoops];
+	for (int i = 0; i < numLoops; i++) {
+	    Test_arraysC data;
+	    fillArraysWithRandomValues<Test_arraysC>(&data);
+	    dataList[i] = data;
+	    REQUIRE(controller->putSample_arrays(&data) == SAL__OK);
+	}
+
+	Test_arraysC *expected_data = &(dataList[numLoops - 1]);
+	Test_arraysC data;
+	remote->getSample_arrays(&data);
+	arraysEqual<Test_arraysC>(expected_data, &data);
+
+	// at this point there should be nothing on the queue
+	REQUIRE(remote->getNextSample_arrays(&data) == SAL__NO_UPDATES);
+	REQUIRE(remote->getSample_arrays(&data) == SAL__NO_UPDATES);
+    }
+
+    /**
+     * Test that get newest after get oldest gets the newest value.
+     * <p>
+     * This tests DM-18491.
+     */
+    SECTION("EvtGetNewestAfterGetOldest") {
+	remote->salEventSub((char *)"Test_logevent_arrays");
+	controller->salEventPub((char *)"Test_logevent_arrays");
+
+	getNewestAfterGetOldest<Test_logevent_arraysC>(
+		std::bind(&SAL_Test::logEvent_arrays, controller, _1, 1),
+		std::bind(&SAL_Test::getNextSample_logevent_arrays, remote, _1),
+		std::bind(&SAL_Test::getSample_logevent_arrays, remote, _1));
+	getNewestAfterGetOldest<Test_logevent_arraysC>(
+		std::bind(&SAL_Test::logEvent_arrays, controller, _1, 1),
+		std::bind(&SAL_Test::getSample_logevent_arrays, remote, _1),
+		std::bind(&SAL_Test::getSample_logevent_arrays, remote, _1), true);
+    }
+
+    /**
+     * Test that get newest after get oldest gets the newest value.
+     * <p>
+     * This tests DM-18491.
+     */
+    SECTION("TelGetNewestAfterGetOldest") {
+	remote->salTelemetrySub((char *)"Test_arrays");
+	controller->salTelemetryPub((char *)"Test_arrays");
+
+	getNewestAfterGetOldest<Test_arraysC>(std::bind(&SAL_Test::putSample_arrays, controller, _1),
+					      std::bind(&SAL_Test::getNextSample_arrays, remote, _1),
+					      std::bind(&SAL_Test::getSample_arrays, remote, _1));
+    }
+
+    SECTION("TelGetNewestAfterGetOldest") {
+	remote->salTelemetrySub((char *)"Test_arrays");
+	controller->salTelemetryPub((char *)"Test_arrays");
+
+	getNewestAfterGetOldest<Test_arraysC>(std::bind(&SAL_Test::putSample_arrays, controller, _1),
+					      std::bind(&SAL_Test::getNextSample_arrays, remote, _1),
+					      std::bind(&SAL_Test::getSample_arrays, remote, _1));
+    }
+
+    /**
+     * Test that a late joiner can can read the most recent event using getNextSample.
+     * <p>
+     * Only one value is retrievable.
```

Comment:
This turns out to be wrong (as your test shows). It gets all historical data.

Please consider expanding the test to read all late-joiner data (and possibly write some non-late-joiner data after constructing the subscriber). Here's the updated Python test, after I fixed the error that made it look like there was just the one item of data:
```
    def check_evt_late_joiner_data(self, read_func, read_gets_oldest):
        """Check event late-joiner data, using the logevent_arrays topic.

        Parameters
        ----------
        read_func : callable
            Function to read data, e.g. self.remote.getEvent_arrays
        read_gets_oldest : bool
            Does the read function get the oldest data?
            Set true for getNextSample and getEvent.
            Set false for getSample.
        """
        self.controller.salEventPub("Test_logevent_arrays")
        time.sleep(STD_SLEEP)

        nhist = 5

        data_list = []
        # Write late-joiner samples (no subscriber yet)
        for i in range(nhist):
            data = SALPY_Test.Test_logevent_arraysC()
            self.test_data.set_random_arrays(data)
            data_list.append(data)
            retcode = self.controller.logEvent_arrays(data, 1)
            time.sleep(STD_SLEEP)
            self.assertEqual(retcode, SALPY_Test.SAL__OK)

        self.remote.salEventSub("Test_logevent_arrays")
        time.sleep(STD_SLEEP)

        if read_gets_oldest:
            # Write new samples
            for i in range(3):
                data = SALPY_Test.Test_logevent_arraysC()
                self.test_data.set_random_arrays(data)
                data_list.append(data)
                retcode = self.controller.logEvent_arrays(data, 1)
                time.sleep(STD_SLEEP)
                self.assertEqual(retcode, SALPY_Test.SAL__OK)

            # getEvent reads the oldest message first, so we should see
            # all historical samples, followed by new samples.
            for expected_data in data_list:
                data = SALPY_Test.Test_logevent_arraysC()
                retcode = read_func(data)
                self.assertEqual(retcode, SALPY_Test.SAL__OK)
                self.test_data.assert_arrays_equal(data, expected_data)
        else:
            expected_data = data_list[-1]
            retcode = read_func(data)
            self.assertEqual(retcode, SALPY_Test.SAL__OK)
            self.test_data.assert_arrays_equal(data, expected_data)

        data = SALPY_Test.Test_logevent_arraysC()
        retcode = read_func(data)
        self.assertEqual(retcode, SALPY_Test.SAL__NO_UPDATES)

    def test_evt_late_joiner_getEvent(self):
        """Test that a late joiner can read historical data using
        getEvent.
        """
        self.check_evt_late_joiner_data(
            read_func=self.remote.getEvent_arrays, read_gets_oldest=True
        )

    def test_evt_late_joiner_getNextSample(self):
        """Test that a late joiner can read historical data using
        getNextSample.
        """
        self.check_evt_late_joiner_data(
            read_func=self.remote.getNextSample_logevent_arrays, read_gets_oldest=True
        )

    def test_evt_late_joiner_getSample(self):
        """Test that a late joiner can read one item of historical data using
        getEvent.
        """
        self.check_evt_late_joiner_data(
            read_func=self.remote.getSample_logevent_arrays, read_gets_oldest=False
        )
```

Reply:
Added. Thanks for this.